### ISSUE TITLE
Parallel Tasks Fail

### ISSUE THREAD

**Issue by nsadawi** [spotify/luigi#2286]
Hi there,
I am using luigi on a kubernetes cluster and I am trying to run 500 parallel jobs (or more) but it freezes and never completes, with some tasks showing as failed in the GUI.
When I click on a failed task to check the error message, I get an empty error box.
I am trying to run the following workflow:

`
# export PYTHONPATH=./ 
# luigi --module many_jobs ManyJobs --scheduler-host luigi.default --workers 500

import luigi
from luigi.contrib.kubernetes import KubernetesJobTask 

class PerlPi(KubernetesJobTask):
    
    name = "pi"
    index = luigi.Parameter()
    max_retrials = 3
    
    @property
    def spec_schema(self): 
        return {
            "containers": [
                {
                    "name": "pi",
                    "image": "perl",
                    "command": ["sh", "-c", "perl -Mbignum=bpi -wle 'print bpi(2000)' > /work/pi"+str(self.index)],
                    "volumeMounts": [{
                        "mountPath": "/work",
                        "name": "shared-volume",
                        "subPath": "jupyter/pi"
                    }]
                }
            ],
            "volumes": [{
                "name": "shared-volume",
                "persistentVolumeClaim": {
                    "claimName": "galaxy-pvc"
                 }
            }]
        }
    
    def output(self):
        target = "pi/pi"+str(self.index)
        return luigi.LocalTarget(target)
    
class ManyJobs(luigi.WrapperTask):
    def requires(self):
        # Generate 500 PerPi jobs  (0, 1 ... 499)
        for i in range(1000):
            yield PerlPi(index=i)
`

Could you please assist?

Many thanks,
Noureddin

**Comment by nsadawi** [comment]
Another issue is that although I ask luigi to run 500 tasks in parallel, only between 200-300 get started and the rest are pending
Best,
Noureddin

**Comment by henryrizzi** [comment]
What kind of cluster are you running this on? In the past I know that I've had issues running with larger numbers of workers when working in a Kubernetes cluster, so I'd try and run with `--workers` set to somewhere close to the amount of CPU cores available

**Comment by nsadawi** [comment]
Hi @henryrizzi ,
it's not clear to me what you mean by kind of cluster, could you please clarify?
My cluster is on the cloud (OpenStack) and it was created using the wonderful KubeNow stack:
https://github.com/kubenow/KubeNow
and
https://github.com/phnmnl/cloud-deploy-kubenow

Regarding the number of CPUs, 500 is ~ 50% of what we have in the cluster
The weird thing is that it works nicely with 100, 200, 300 workers, but when I try 400 and above, the problem starts to appear

Many thanks

**Comment by nsadawi** [comment]
@henryrizzi another point is that I can manually run as many tasks (or pods) as I want on our k8s cluster, this is why I believe it's a luigi problem

@mcapuccini .. could you please comment?

**Comment by nsadawi** [comment]
To add to this, although when I use 500 workers some tasks start and some others are pending, in the end it all hangs and luigi tells me the tasks that have started failed. It's strange that in reality they work as expected and the files they're supposed to create are there.
Please see attached screenshots
luigi terminal
https://drive.google.com/open?id=1BbTpT36h0ISkz3d3anxKJ6bcfWDm0N8K

luigi gui
https://drive.google.com/open?id=1XoKMdUi3dfkgCpOrK4EYlys_3MxifG-z

luigi failed tasks
https://drive.google.com/open?id=1uaP-KylBElSVoTcORLv25f1dSWgd2IS1

Thanks

**Comment by mcapuccini** [comment]
Hi. We need to isolate the problem. Can you try to run a simple Luigi workflow with 500 workers? No Kubernetes tasks, just luigi tasks.

**Comment by mcapuccini** [comment]
I think that the problem is that Luigi breaks down since a process is spawned for each workers, so for 500 workers there are probably network timeouts and everything starts to break down. Luigi is not meant to massively parallelize jobs, in my opinion this thread can be closed. Maybe someone from the Spotify community can further confirm this.

**Comment by nsadawi** [comment]
From luigi's online docs:
`The downside is that Luigi doesn’t give you scalability for free. In practice this is not a problem until you start running thousands of tasks.`
and
`The downside is that Luigi doesn’t give you scalability for free. In practice this is not a problem until you start running thousands of tasks.`

We're only doing 100's so we still hope to make it heppen

@henryrizzi What do you think?

**Comment by Tarrasch** [comment]
@nsadawi, ah the docs isn't clear. You can have about tens (I don't know, up to 30? 50?) of parallely executing workers in one luigi invocation. But there can be 1000s of tasks (nodes) in the dependency tree, they will be queued and eventually executed, but I wouldn't count on that they all can run at the same time from the same machine. Remember, Luigi isn't a hadoop/spark replacement. :)

**Comment by nsadawi** [comment]
@Tarrasch The way our stuff is working is that each worker is essentially a kubernetes pod and it only runs one task so no much worry about dependencies. When I use this command:
`$> luigi --module many_jobs ManyJobs --scheduler-host luigi.default --workers 500`
~350 workers are initialised and the rest are pending.
After that it all freezes and nothing happens
I don't think we have any RAM or CPU issues as we're monitoring their usage.
Any suggestions please?
Thanks a million

**Comment by Tarrasch** [comment]
If you just can't use less workers. Can you do `--workers 50` but on 10 different machines (all against the same central scheduler).

**Comment by nsadawi** [comment]
Choosing the machines to host the workers (i.e. pods) is up to kubernetes so that's not a problem for us at all. I can nicely run 10, 50, 100, 200, 300 workers .. more than that it hangs
My target is 500 as I'm doing a scalability study

**Comment by mcapuccini** [comment]
@nsadawi the workaround that @Tarrasch suggested would work. I'll contact you on our Slack channel to give you more help.

**Comment by nsadawi** [comment]
@mcapuccini yes, that'd be great, thanks!

**Comment by nsadawi** [comment]
@Tarrasch one question .. luigi scheduler uses all available cpus by default?
I monitor the VM it's running on and it's not showing me that the cpus are busy when submitting hundreds of workers 
Any clues please?
